### PR TITLE
Sets up simulated GCS bucket directory browsing; fills in blob metadata

### PR DESCRIPTION
**Pull request by etanshaul**
This PR implements directory browsing in the GCS explorer. It also fills in the remaining blob metadata.

In the backend, there is no concept of "directories" within a GCS bucket. Everything is flattened and a blob just has a name and metadata. Therefore, simulation of directories was necessary to create the desired UX (a file browser-like experience). For well written writeup on this, see https://github.com/csells/BucketHelperSample

Note: this PR does not add any visual polish to the browser. Subsequent PRs will address look and feel / icons. Also remaining is implementation of the toolbar (with upload/download/refresh) and a breadcrumb navigator.

File browser:
image

Double clicking a directory, reloads the display and shows the contents of the clicked directory:
image

### CODE REVIEW

**Comment by etanshaul** [comment]
@nkibler, @patflynn you can start taking a look. Still need to update and add more tests.

**Comment by etanshaul** [comment]
Note, for this to work, https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1624 (or some similar solution) will need to be merged in first.

**Comment by etanshaul** [comment]
@nkibler @patflynn this is ready for review. PTAL.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java [hunk 1]
```diff
@@ -17,36 +17,69 @@
...
 import javax.swing.JTable;
-import javax.swing.table.DefaultTableModel;
 import org.jetbrains.annotations.NotNull;
 
 /** Defines the Google Cloud Storage bucket content browsing UI panel. */
-// TODO(eshaul) work in progress
+// TODO(eshaul) implement toolbar panel and breadcrumb browser
 final class GcsBucketContentEditorPanel {
 
+  private Bucket bucket;
```
Comment (nkibler) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java]: final?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java [hunk 2]
```diff
@@ -17,36 +17,69 @@
...
   }
 
+  private void updateTableModel(String prefix) {
+    tableModel.setRowCount(0);
+    tableModel.setDataVector(getBlobsAtLevel(prefix), prefix);
+    tableModel.fireTableDataChanged();
+  }
+
+  // TODO(eshaul) should be done asynchronously and show loader in the UI
+  private List<Blob> getBlobsAtLevel(String prefix) {
```
Comment (nkibler) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java]: I'm not sure what "level" refers to here. Maybe this should be renamed to `getBlobsForPrefix()`?

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java [hunk 3]
```diff
@@ -36,51 +40,97 @@
...
+  @Rule public final TimeZoneRule timeZoneRule = new TimeZoneRule(TimeZone.getTimeZone("GMT"));
 
   private GcsBucketContentEditorPanel editorPanel;
   private GcsBucketVirtualFile bucketVirtualFile;
-  @Mock private Blob blob;
+
+  @Mock private Blob directoryBlob;
+  @Mock private Blob binaryBlob;
+
+  private BiMap<Integer, String> indexToColName =
```
Comment (nkibler) [inline google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: Can these be `static final`?

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java [hunk 4]
```diff
@@ -36,51 +40,97 @@
...
+      HashBiMap.create(ImmutableMap.of(0, "Name", 1, "Size", 2, "Type", 3, "Last Modified"));
+  private Map<String, Integer> colNameToIndex = indexToColName.inverse();
 
   @Before
   public void setUp() {
     bucketVirtualFile = GcsTestUtils.createVirtualFileWithBucketMocks();
+
+    when(directoryBlob.isDirectory()).thenReturn(true);
+    when(directoryBlob.getName()).thenReturn("my_directory");
```
Comment (nkibler) [inline google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: Maybe pull these into static constants so writing tests below don't require you to read the `setUp()` method every time to remember what these values are?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java (hunk 2 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java]: I updated the name to `getBlobsStartingWith` and added javadoc to explain whats going on. Let me know if its more clear now.

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java (hunk 3 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: done

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java (hunk 4 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: done

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java (hunk 1 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java]: done

File: common-test-lib/src/main/java/com/google/cloud/tools/intellij/testing/TimeZoneRule.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+package com.google.cloud.tools.intellij.testing;
+
+import java.util.TimeZone;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/** A custom {@link TestRule} for setting a time zone in unit tests. */
+public class TimeZoneRule implements TestRule {
```
[the remark this answers is not in the document]
Comment (etanshaul) [reply common-test-lib/src/main/java/com/google/cloud/tools/intellij/testing/TimeZoneRule.java]: done

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBlobTableModel.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  private static final List<String> TABLE_COL_HEADER =
+      Arrays.asList("Name", "Size", "Type", "Last Modified");
+
+  @Override
+  public boolean isCellEditable(int row, int column) {
+    return false;
+  }
+
+  Blob getBlobAt(int index) {
+    return blobs.get(index);
```
[the remark this answers is not in the document]
Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBlobTableModel.java]: hmm. if blobs is empty this method will still throw an IndexOutOfBounds excpetion. What about throwing a runtime exception if blobs is null?

**Comment by etanshaul** [comment]
@nkibler ptal. Addressed your comments. Also added some more tests.

Regarding your comment about `getBlobAt` throwing an NPE, I updated to throw a RuntimeException if blobs are null. let me know what you think.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java [hunk 7]
```diff
@@ -60,22 +60,32 @@ public void mouseClicked(MouseEvent event) {
...
-  private void updateTableModel(String prefix) {
+  void updateTableModel(String prefix) {
     tableModel.setRowCount(0);
-    tableModel.setDataVector(getBlobsAtLevel(prefix), prefix);
+    tableModel.setDataVector(getBlobsStartingWith(prefix), prefix);
     tableModel.fireTableDataChanged();
   }
 
+  /**
+   * In the GCS backend there are no directories, just blobs with names. Implicit in the name,
```
Comment (nkibler) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java]: nit: This seems like a supporting paragraph that should exist under a summary fragment for this method. Consider adding a summary fragment like:

```java
/**
 * Returns the list of {@link Blob Blobs} that start with the given prefix.
 *
 * <p>In the GCS backend...
 */
```

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java [hunk 8]
```diff
@@ -47,23 +47,32 @@
 
   @Mock private Blob directoryBlob;
   @Mock private Blob binaryBlob;
+  @Mock private Blob nestedBlob;
 
-  private BiMap<Integer, String> indexToColName =
+  private static final String DIR_NAME = "my_directory";
```
Comment (nkibler) [inline google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: nit: Can you move these static variables above the instance variables to maintain a logical ordering?

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java [hunk 9]
```diff
@@ -47,23 +47,32 @@
...
+  private static final String BLOB_NAME = "my_blob.zip";
+  private static final String BLOB_CONTENT_TYPE = "application/zip";
+  private static final String NESTED_BLOB_FULL_NAME = "dir1/dir2/nested_blob.zip";
+  private static final String NESTED_BLOB_NAME = "nested_blob.zip";
+
+  private static final BiMap<Integer, String> indexToColName =
```
Comment (nkibler) [inline google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: nit: Use CONSTANT_CASE for these, now that they're constants.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java (hunk 7 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java]: done

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java (hunk 8 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: done

File: google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java (hunk 9 above)
Comment (etanshaul) [reply google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanelTest.java]: done

**Pull request merged**